File: src/main/java/puzzles/year24/day4/Part2.java
package puzzles.year24.day4;

import util.Puzzle;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Part2 implements Puzzle<Integer>  {
    @Override
    public Integer getAnswer(String inputString) {
        char[][] grid = Arrays.stream(inputString.split(System.lineSeparator())).map(String::toCharArray).toArray(char[][]::new);
        int result = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                char c = grid[i][j];

                if (c != 'A') continue;
                if(isX(grid, i, j)) {
                    result++;
                }
            }
        }

        return result;
    }

    private boolean isX(char[][] grid, int i, int j) {
        return checkLeft(grid, i, j) && checkRight(grid, i, j);
    }

    private boolean checkLeft(char[][] grid, int i, int j) {
        return (step(grid, new int[] {1, 1}, i, j) == 'M' && step(grid, new int[] {-1, -1}, i, j) == 'S') || (step(grid, new int[] {1, 1}, i, j) == 'S' && step(grid, new int[] {-1, -1}, i, j) == 'M');
    }

    private boolean checkRight(char[][] grid, int i, int j) {
        return (step(grid, new int[] {1, -1}, i, j) == 'M' && step(grid, new int[] {-1, 1}, i, j) == 'S') || (step(grid, new int[] {1, -1}, i, j) == 'S' && step(grid, new int[] {-1, 1}, i, j) == 'M');
    }

    private char step(char[][] grid, int[] direction, int i, int j) {
        int iMax = grid.length;
        int jMax = grid[0].length;

        int iIndex = i + direction[0];
        int jIndex = j + direction[1];

        if (iIndex >= iMax || jIndex >= jMax || iIndex < 0 || jIndex < 0) {
            return '~';
        } else {
            return grid[iIndex][jIndex];
        }
    }
}

File: src/main/java/puzzles/year25/day7/Part2.kt
package puzzles.year25.day7

import util.Puzzle

class Part2 : Puzzle<Long?> {
    override fun getAnswer(inputString: String): Long {
        val lines = inputString.trim().lines().map { it.toCharArray() }
        val waysToGetHere: Array<Array<Long>> = Array(lines.size + 1) {
            Array(lines[0].size) { 0 }
        }

        waysToGetHere[0][lines[0].indexOf('S')] = 1

        for (i in lines.indices) {
            for (j in lines[i].indices) {
                val c = lines[i][j]
                val waysToGetC = waysToGetHere[i][j]
                if (waysToGetC == 0L) continue
                if (c == '^') {
                    waysToGetHere[i + 1][j - 1] += waysToGetC
                    waysToGetHere[i + 1][j + 1] += waysToGetC
                } else {
                    waysToGetHere[i + 1][j] += waysToGetC
                }
            }
        }

        return waysToGetHere.last().sumOf { it }
    }
}
